strucutre improvement

mkangia opened the pull request:
Follow up on https://github.com/dimagi/commcare-hq/pull/21516

Basically picked from these two feedbacks
https://github.com/dimagi/commcare-hq/pull/21516#pullrequestreview-149742103 by Daniel
https://github.com/dimagi/commcare-hq/pull/21516#discussion_r213485235 by Ethan

Changes made, also in the commit message
- make PoFileGenerator behave as a context manager
- revert TransifexPoFileGenerator to have a method that does the major task[was not needed and was removed eventually]
- rename TransifexPoFileGenerator to AppTranslationsGenerator

Why?
Context manager was definitely the way to go. Got all clean up code at one place.
I reverted the change i made earlier for TransfiexPoFileGenerator, specifically [here](https://github.com/dimagi/commcare-hq/pull/21723/files#diff-389860842dade5566b99beb0b7021f2eR192) and thought of bringing all its operations in a single method as suggested by Ethan.
After the changes the functionality broke. so then i made `Transifex` co-ordinate directly with `PoFileGenerator` and renamed `TransfiexPoFileGenerator` as `AppTranslationsGenerator` as it was now just generating the content for the translations files.

Thanks for suggesting this.

mkangia committed:
strucutre improvement

- make PoFileGenerator behave as a context manager
- revert TransifexPoFileGenerator to have a method that does the major task
- rename TransifexPoFileGenerator to AppTranslationsGenerator

millerdev commented on corehq/apps/translations/views.py:
```diff
@@ -101,11 +101,11 @@ def _generate_translations_for_po(self, worksheet):
-        po_file_generator = PoFileGenerator(translations, {})
-        return po_file_generator
+        with PoFileGenerator(translations, {}) as po_file_generator:
+            return po_file_generator.generated_files[0][1]
```
The generated file will be cleaned up before it is returned. To get around that you could read the content in this method before cleanup happens:
```py
def _generate_po_content(self, worksheet):
    """Extract translations from worksheet and convert to PO file content

    :return: UTF-8 string of PO file content
    """
    translations = self._generate_translations_for_po(worksheet)
    with PoFileGenerator(translations, {}) as po_file_generator:
        file_path = po_file_generator.generated_files[0][1]
        with open(file_path, 'r', encoding="utf-8") as f:
            return f.read()
```
the calling code would look like this
```py
content = self._generate_po_content(worksheet)
```

mkangia replied:
Aha! thanks.. missed this.

millerdev commented on corehq/apps/translations/views.py:
```diff
@@ -101,11 +101,11 @@ def _generate_translations_for_po(self, worksheet):
-        po_file_generator = PoFileGenerator(translations, {})
-        return po_file_generator
+        with PoFileGenerator(translations, {}) as po_file_generator:
+            return po_file_generator.generated_files[0][1]
```
Unrelated question: use of index constants in `.generated_files[0][1]` is a bit magic-y. Why only process the first file? If there is only ever one, could `.generated_files` be refactored to be `.generated_file`?

Finally, named tuple would be nice for readability here, making it `.generated_files[0].path` with the named tuple defined as
```py
from collections import namedtuple

PoFileInfo = namedtuple("PoFileInfo", "name path")
```
and

https://github.com/dimagi/commcare-hq/blob/7d42f35fb4d4305ec1c75131f6808639cb74b1ef/corehq/apps/app_manager/app_translations/generators.py#L228

would change to
```py
self.generated_files.append(PoFileInfo(file_name, temp_file.name))
```

mkangia replied:
So,
`PoFileGenerator` has the ability to generate multiple files. This django view however just picks the first sheet of the excel file uploaded because of the workflow after this process explained below. One can definitely modify this to process all sheets in the uploaded excel, generate a bunch of PO files, zip them and return that to the user. But currently there is not a use case with multiple sheets.

A little more context,
The way its used is that one needs to convert an excel sheet(not workbook) to a PO file and then upload it to Transifex manually via their interface. Till now we have always done one file at a time. I am open to making it do what i said above, just didn't see the use for stretching it so far yet.

millerdev replied:
Am I correct in understanding that `po_file_generator.generated_files` will contain files for every sheet even though only the first one will be used? Seems like it's doing more work than is necessary. I suppose that's a reasonable implementation detail if performance of processing all sheets is not a problem.

The namedtuple would still be nice.

mkangia replied:
> Am I understanding correctly that po_file_generator.generated_files will contain a file for every sheet even though only the first one will be used?

The view picks only the first sheet and gives that to `PoFileGenerator`, as implemented [here](https://github.com/dimagi/commcare-hq/blob/15f567011ad02213351f9bc7d14ca907c35e1fc5/corehq/apps/translations/views.py#L131-L130).

> The namedtuple would still be nice.

Yup! made that change. Could not push because of internet :) .You should be able see the change shortly.

millerdev reviewed (commented):
Looking good. Few comments.

esoergel commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -212,9 +198,15 @@ def __init__(self, translations, metadata):
             exc_info = sys.exc_info()
-            self.cleanup()
+            self._cleanup()
             six.reraise(*exc_info)
```
You can restructure this context manager a bit to move the work out of the constructor.  The nice thing about that approach is that you could then drop this try/except/cleanup stuff, since `__exit__` will take care of that.

Specifically, you could drop `self.generated_files` entirely and make `generate_translation_files` public and call that in the client code (I'd save `translations` and `metadata` on the class so you don't need to pass them twice).  You could memoize `generate_translation_files` or make it a memoized property if it needs to be accessed more than once.

mkangia replied:
Thanks @esoergel . Made all the changes.

mkangia replied:
sorry, made one comment earlier. Looked like it didnt reach Github.

> You could memoize generate_translation_files or make it a memoized property if it needs to be accessed more than once.

Considering the changes done now, i don't see why one would need to access generated files more than once. Plus they can't anyway since it gets cleaned up. So don't think making it as `memoized` property is needed.

mkangia committed:
return content instead of file

The PoFileGenerator would clean the file as soon as its context is lost, and hence the file wont be available to read later.

mkangia committed:
refactor PoFileGenerator

expose generate_translation_files and get rid of complexity in __init__

mkangia committed:
use POFileInfo for better readability

mkangia committed:
update documentation

esoergel reviewed (approved):
Very nice!

millerdev commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -190,37 +189,25 @@ def _metadata(self):
 
-class PoFileGenerator(object):
-    def __init__(self, translations, metadata):
-        self.generated_files = list()  # list of tuples (filename, filepath)
```
Removing this will break `self._cleanup()`. Recommendation: convert this class (`PoFileGenerator`) into a `contextmanager`-decorated function:
```py
from contextlib import contextmanager

@contextmanager
def generate_po_files(translations, metadata):
    """Generate PO files from translations

    Usage:

        with generate_translation_files(trans, meta) as files:
            for name, path in files:
                # do stuff with name, path
        # generated files are cleaned up (deleted) on with block exit

    :param translations: dict of translations:
    `{filename: [list of translations], ...}`
    :param metadata: PO file metadata.
    :yields: (context object) list of `POFileInfo` objects.
    """
    generated_files = []
    try:
        for file_name in translations:
            sheet_translations = translations[file_name]
            po = polib.POFile()
            po.check_for_duplicates = False
            po.metadata = metadata
            for translation in sheet_translations:
                source = translation.key
                if source:
                    entry = polib.POEntry(
                        msgid=translation.key,
                        msgstr=translation.translation or '',
                        occurrences=translation.occurrences,
                        msgctxt=translation.msgctxt
                    )
                    po.append(entry)
            temp_file = tempfile.NamedTemporaryFile(delete=False)
            po.save(temp_file.name)
            generated_files.append(POFileInfo(file_name, temp_file.name))
        yield generated_files
    finally:
        for resource_name, filepath in generated_files:
            if os.path.exists(filepath):
                os.remove(filepath)
```

esoergel replied:
Ah right - that'd work.  You could also keep around a reference to the files as a private variable.

mkangia replied:
Oh! sorry for the oversight. Got carried away with removing code. Adding as a private variable.

I would still like to keep this as a class.. There is quite a lot happening here to make it just a method and i would like to keep it grouped together, but there is definitely something to learn from that example. Thanks @millerdev

millerdev commented on corehq/apps/translations/views.py:
```diff
@@ -106,7 +106,7 @@ def _generate_po_content(self, worksheet):
         with PoFileGenerator(translations, {}) as po_file_generator:
             generated_files = po_file_generator.generate_translation_files()
-            with open(generated_files[0][1], 'r', encoding="utf-8") as f:
+            with open(generated_files[0].file_path, 'r', encoding="utf-8") as f:
```
Should say `generated_files[0].path`

mkangia committed:
fix name

mkangia committed:
add back generated_files to not break _cleanup

mkangia commented:
I tested this earlier, but now there are some changes after that so added the label to test it before merging, once approved.
